runtime/parsing_functions.cpp

```diff
@@ -0,0 +1,325 @@
+      while (ini_string_copy[i] != '\"' && ini_string_copy[i] != '\'') {
+        ini_entry.push_back(ini_string_copy[i]);
+        i++;
+      }
+    }
+
+    if (ini_string_copy[i] == ' ' || ini_string_copy[i] == '\n' || ini_string_copy[i] == '\0') {
+      if (isINISection(ini_entry)) {
+        if (process_sections && !ini_section.empty()) {
+          res.set_value(ini_section, section);
+          section.clear();
+        }
+
+        ini_section = get_ini_section(ini_entry);
+      } else {
+        string ini_var = get_ini_var(ini_entry);
+        string ini_val = get_ini_val(ini_entry);
+
+        if (!isINIVar(ini_var) && !isINIVal(ini_val)) {
+          php_warning("Invalid ini string format %s", ini_entry.c_str());
+          return {};
+        }
+
+        if (!ini_var.empty()) {
+          switch (scanner_mode) {
+            case INI_SCANNER_NORMAL:
+              if (isINIBoolVal(ini_val)) {
+                section.set_value(ini_var, cast_str_to_bool(ini_val) ? string("1") : string(""));
+              } else if (isINIStrVal(ini_val)) {
+                section.set_value(ini_var, clearQuotes(ini_val));
+              }
+
+              break;
+            case INI_SCANNER_RAW:
+              if (isINIStrVal(ini_val)) {
+                section.set_value(ini_var, clearDoubleQuotes(ini_val));
+              } else {
+                section.set_value(ini_var, ini_val);
+              }
+
+              break;
+            case INI_SCANNER_TYPED:
+              if (ini_val.is_int()) {
+                section.set_value(ini_var, ini_val.to_int());
+              } else if (isINIFloatVal(ini_val)) {
+                section.set_value(ini_var, ini_val.to_float());
+              } else if (isINIBoolVal(ini_val)) {
+                section.set_value(ini_var, cast_str_to_bool(ini_val));
+              } else if (isINIStrVal(ini_val)) {
+                section.set_value(ini_var, clearQuotes(ini_val));
+              }
+
+              break;
+          }
+        }
+      }
+
+      ini_entry = {};
+    } else {
+      ini_entry.push_back(ini_string_copy[i]);
+    }
+  }
+
+  if (process_sections) {
+    if (!ini_section.empty()) {
+      res.set_value(ini_section, section);
+    }
+  } else {
+    return section;
+  }
+
+  return res;
+}
+
+
+
+array<mixed> f$parse_ini_file(const string &filename, bool process_sections, int scanner_mode)
+{
+  if (filename.empty()) {
+    php_warning("Filename cannot be empty");
+    return {};
+  }
+
+  std::ifstream ifs(filename.c_str());
+
+  if (!ifs.is_open()) {
+    php_warning("failed to open stream: %s", filename.c_str());
+    return {};
+  }
+
+  std::string str;
+  std::string ini_string;
+
+  while (std::getline(ifs, str)) {
+    if (!str.empty()) {
+      ini_string.append(str);
+      ini_string.append(" ");
+    }
+  }
+
+  if (!str.empty()) {
+    ini_string.append(str);
+  }
+
+  ifs.close();
+
+  return f$parse_ini_string(string(ini_string.c_str()), process_sections, scanner_mode);
```

Comment:
Using `std::string` and `std::ifstream` this way is unsafe as I explained above. But here you don't need it, just use `f$file_get_contents` to read the file